lib/tiki/torch/publishing/retries.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+module Tiki
+  module Torch
+    module Publishing
+      class Retries
+        # Class to encapsulate failed attempts to publish an event to a topic
+        #
+        # @attr [String] topic_name
+        # @attr [Object] event
+        # @attr [Integer] tries
+        class Entry
+          attr_reader :topic_name, :event, :tries
+
+          # @param [String] topic_name the name of the topic to retry
+          # @param [Object] event the event to retry
+          def initialize(topic_name, event)
+            @topic_name = topic_name
+            @event      = event
+            @tries      = 1
+          end
+
+          # @return [String] A unique identifier for the topic_name + event
+          def fingerprint
+            @fingerprint ||= format '%s:%s',
+                                    @topic_name.to_s,
+                                    (@event.respond_to?(:fingerprint) ? @event.fingerprint : @event.to_s)
+          end
+
+          # Increase the quantity of tries for the entry
+          # This is informational only in that we do not stop retrying at the moment.
+          def increment_tries
+            @tries += 1
+          end
+        end
+
+        include Logging
+
+        class << self
+          # A list of entries that failed to publish. The key is the fingerprint of the event.
+          #
+          # @return [Concurrent::Hash{String=>Tiki::Torch::Publishing::Retries::Entry}]
+          def entries
+            @entries ||= Concurrent::Hash.new
+          end
+
+          # Adds or increments an entry to the list. Runs the callback so the entry can be reported
+          # or removed given the number of retries or other criteria
+          #
+          # @param [String] topic_name the name of the topic to retry
+          # @param [Object] event the event to retry
+          # @return [Tiki::Torch::Publishing::Retries::Entry] the job to retry
+          def add(topic_name, event)
+            entry = Entry.new topic_name, event
+            found = entries[entry.fingerprint]
+            if found
+              found.increment_tries
+              Torch.config.publishing_retry_handler.call entries, found
+            else
+              entries[entry.fingerprint] = entry
+              Torch.config.publishing_retry_handler.call entries, entry
+            end
+          end
+
+          # The callback to deal with the list of entries and the entry that was just added
+          # to that list
+          #
+          # @param [Hash{String=>Tiki::Torch::Publishing::Retries::Entry}] current_entries
+          # @param [Tiki::Torch::Publishing::Retries::Entry] entry
+          # @return [Tiki::Torch::Publishing::Retries::Entry]
+          def default_retry_handler(_current_entries, entry)
+            entry
```

Comment:
Nice. So by default, it does nothing, but somebody could implement queue clearing. That seems simple and configurable.